# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

In [2]:
cat helper.py

import os
import pickle
import torch


SPECIAL_WORDS = {'PADDING': '<PAD>'}


def load_data(path):
    """
    Load Dataset from File
    """
    input_file = os.path.join(path)
    with open(input_file, "r") as f:
        data = f.read()

    return data


def preprocess_and_save_data(dataset_path, token_lookup, create_lookup_tables):
    """
    Preprocess Text Data
    """
    text = load_data(dataset_path)
    
    # Ignore notice, since we don't use it for analysing the data
    text = text[81:]

    token_dict = token_lookup()
    for key, token in token_dict.items():
        text = text.replace(key, ' {} '.format(token))

    text = text.lower()
    text = text.split()

    vocab_to_int, int_to_vocab = create_lookup_tables(text + list(SPECIAL_WORDS.values()))
    int_text = [vocab_to_int[word] for word in text]
    pickle.dump((int_text, vocab_to_int, int_to_vocab, token_dict), open('preprocess.p', 'wb'))


def load_preprocess():
    """
    Load the Preprocessed Training data a

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [5]:
view_line_range = (0, 50)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 50:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [6]:
import problem_unittests as tests

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    sorted_vocab = sorted(set(text))
    
    vocab_to_int = {}
    int_to_vocab = {}
    
    for index,word in enumerate(sorted_vocab):
        vocab_to_int[word] = index
        int_to_vocab[index] = word
    
    # return tuple
    return (vocab_to_int, int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [16]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    punc_to_token = {}
    
    punc_to_token["."] = "||Period||"
    punc_to_token[","] = "||Comma||" 
    punc_to_token[";"] = "||Semicolon||" 
    punc_to_token["!"] = "||Exclamation||" 
    punc_to_token["?"] = "||Questionmark||" 
    punc_to_token["("] = "||Left_Parentheses||" 
    punc_to_token[")"] = "||Right_Parentheses||" 
    punc_to_token["-"] = "||Dash||" 
    punc_to_token["\n"] = "||Return||" 
    punc_to_token['"'] = "||Quotation_mark||" 

    return punc_to_token

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

In [2]:
len(int_to_vocab), len(vocab_to_int)

(21388, 21388)

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')
else:
    print('yay! TRAINING ON GPU!!')

yay! TRAINING ON GPU!!


## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [5]:
from torch.utils.data import TensorDataset, DataLoader
import numpy as np

def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
    feature_size = len(words) - sequence_length
    
    train_x = np.zeros((feature_size, sequence_length),  dtype=int)
    train_y = np.zeros(feature_size)
    
    for index in range(feature_size):
        train_x[index] = words[index:index+sequence_length]
        
        # train_y[index] has to be next value, note that in train_x[index] words[index+sequene_length] was exclusive
        train_y[index] = words[index+sequence_length] 
    
    feature_tensors, target_tensors = np.array(train_x, np.int64), np.array(train_y, np.int64)
    
    feature_tensors, target_tensors = torch.from_numpy(feature_tensors), torch.from_numpy(target_tensors)
    
    tensor_data = TensorDataset(feature_tensors, target_tensors)

    data_loader = DataLoader(tensor_data, batch_size=batch_size, shuffle = True)
    
    # return a dataloader
    return data_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own
example_words = list(range(1,10))
seq_length = 3
example_loader = batch_data(example_words, seq_length, 3)

data_iter = iter(example_loader)
next(data_iter)

[tensor([[ 4,  5,  6],
         [ 1,  2,  3],
         [ 3,  4,  5]]), tensor([ 7,  4,  6])]

### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [6]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 39,  40,  41,  42,  43],
        [ 13,  14,  15,  16,  17],
        [ 10,  11,  12,  13,  14],
        [ 34,  35,  36,  37,  38],
        [ 27,  28,  29,  30,  31],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  8,   9,  10,  11,  12],
        [  7,   8,   9,  10,  11],
        [ 24,  25,  26,  27,  28]])

torch.Size([10])
tensor([ 44,  18,  15,  39,  32,  43,  30,  13,  12,  29])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [45]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # TODO: Implement function
        
        # set class variables
        self.output_size = output_size
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim
        
        # define model layers
        
        # embedding and LSTM layers
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers, dropout=dropout, batch_first=True)
        
        # add dropout layer
        self.dropout = nn.Dropout(0.3)
        
        # add fully connected layer
        self.fc = nn.Linear(hidden_dim, output_size)
        
        
        
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function   
        
        batch_size = nn_input.size(0)
        
        # embedding and lstm output
        embeds = self.embedding(nn_input)
        
#         adding flatten parameters as per traing warning
#         self.lstm.flatten_parameters()
        
        lstm_out, hidden = self.lstm(embeds, hidden)
        
        # now stacking lstm outputs
        
        lstm_out = lstm_out.contiguous().view(-1, self.hidden_dim)
        
        # dropout and fully connected layer
        output = self.dropout(lstm_out)
        output = self.fc(output)
        
        # get the last batch of word scores by shaping the output of the final, fully-connected layer
        # reshape into (batch_size, seq_length, output_size)
        output = output.view(batch_size, -1, self.output_size)
        
        # get last batch
        out = output[:, -1]
        
        # return one batch of output word scores and the hidden state
        return out, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        
        weight = next(self.parameters()).data
        
        # initialize hidden state with zero weights, and move to GPU if available
        
        if train_on_gpu:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())
        
        return hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [46]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param rnn: The PyTorch Module that holds the neural network
    :param optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    
    # move data to GPU, if available

    if train_on_gpu:
        inp, target = inp.cuda(), target.cuda()
    
    # perform backpropagation and optimization
    
    # Creating new variables for the hidden state, otherwise we'd backprop through the entire training history
    hidden = tuple([hidden_state.data for hidden_state in hidden])
    
    # zero accumulated gradients
    rnn.zero_grad()
    
    # get the output from the model
    output, hidden = rnn(inp, hidden)
    
    # calculating loss and doing back prop
    loss = criterion(output, target)
    loss.backward()
    
    # `clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
    nn.utils.clip_grad_norm_(rnn.parameters(), 5)
    
    optimizer.step()
    
    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_rnn` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [47]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [29]:
len(int_to_vocab), len(vocab_to_int)

(21388, 21388)

In [43]:
# Data params
# Sequence Length
sequence_length =  10 # of words in a sequence
# Batch Size
###########################  FIRST, SECOND EXPERIMENT  ####################################################

# batch_size = 128

###########################  THIRD EXPERIMENT  ############################################################

batch_size = 120

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [ ]:
len(train_loader.dataset), len(train_loader.dataset) // batch_size

In [52]:
# Training parameters
# Number of Epochs

###########################  FIRST EXPERIMENT  ############################################################
# num_epochs = 15

# now training for few epochs only with new hyperparameters as I have already trained it for 15 epochs,
# as I am going to use already trained model using previous trained result using checkpoint,
# the loss will resume from past experiment, hence I decided to train for 5 epochs only this time with the
# objective to reduce loss below 3.5 as required
###########################  SECOND EXPERIMENT  ############################################################
# num_epochs = 5

###########################  THIRD EXPERIMENT  ############################################################
# num_epochs = 10

# again training 5 more epochs as loss are near to what is expected
num_epochs = 5


# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size

# Embedding Dimension
###########################  FIRST, THIRD EXPERIMENT  ####################################################

embedding_dim = 256

###########################  SECOND EXPERIMENT  ############################################################
# changing embed dimension from 256 to 150 after initial experiment
# embedding_dim = 150

# Hidden Dimension
hidden_dim = 512

# Number of RNN Layers
###########################  FIRST EXPERIMENT  ############################################################
# n_layers = 3

###########################  SECOND, THIRD EXPERIMENT  ####################################################
# changing num of layers from 3 to 2 after initial experiment
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 500

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

### First experiment

In [14]:
# """
# DON'T MODIFY ANYTHING IN THIS CELL
# """

# # create model and move to gpu if available
# rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
# if train_on_gpu:
#     rnn.cuda()

# # defining loss and optimization functions for training
# optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
# criterion = nn.CrossEntropyLoss()

# # training the model
# trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# # saving the trained model
# helper.save_model('./save/trained_rnn', trained_rnn)
# print('Model Trained and Saved')

Training for 15 epoch(s)...
Epoch:    1/15    Loss: 5.565617721557617

Epoch:    1/15    Loss: 4.949369384288788

Epoch:    1/15    Loss: 4.751260939598083

Epoch:    1/15    Loss: 4.647412904262543

Epoch:    1/15    Loss: 4.5620755457878115

Epoch:    1/15    Loss: 4.49979297542572

Epoch:    1/15    Loss: 4.453883576393127

Epoch:    1/15    Loss: 4.416010299682617

Epoch:    1/15    Loss: 4.390609561443329

Epoch:    1/15    Loss: 4.359482147216797

Epoch:    1/15    Loss: 4.331617206573486

Epoch:    1/15    Loss: 4.313855944156646

Epoch:    1/15    Loss: 4.306440577983857

Epoch:    2/15    Loss: 4.212414000917637

Epoch:    2/15    Loss: 4.142964904785156

Epoch:    2/15    Loss: 4.130095574855805

Epoch:    2/15    Loss: 4.133323279380798

Epoch:    2/15    Loss: 4.12448962020874

Epoch:    2/15    Loss: 4.111533119678497

Epoch:    2/15    Loss: 4.108568336009979

Epoch:    2/15    Loss: 4.09740874004364

Epoch:    2/15    Loss: 4.08652193069458

Epoch:    2/15    Loss: 4.075

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


Model Trained and Saved


### As the initital hyperparameters didn't gave praising results, the loss is not decreasing well, so I am going to change number of layers and embedding size this time and train again

### Note: here I am passing `trained RNN` from past experiment, loaded using below Checkpoint. In this step, I have runned below code block of `Checkpoint` first to load trained model so I can resume with new hyperparameters without having to train from beginning. Thus this enables us to resume loss from last traning batch.

#### SECOND EXPERIMENT

In [49]:
# """
# DON'T MODIFY ANYTHING IN THIS CELL
# """

# # create model and move to gpu if available
# rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
# if train_on_gpu:
#     rnn.cuda()

# # defining loss and optimization functions for training
# optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
# criterion = nn.CrossEntropyLoss()

# # training the model
# #######################################################################################
# trained_rnn = train_rnn(trained_rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# # saving the trained model
# helper.save_model('./save/trained_rnn', trained_rnn)
# print('Model Trained and Saved')

### Again the loss was was converging so I interrupted the traning, now I am going to train from the beginning with new hyperparameters

### THIRD EXPERIMENT

In [48]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 10 epoch(s)...
Epoch:    1/10    Loss: 5.332671521186828

Epoch:    1/10    Loss: 4.792315729141236

Epoch:    1/10    Loss: 4.634694821357727

Epoch:    1/10    Loss: 4.524869334697724

Epoch:    1/10    Loss: 4.425402616977691

Epoch:    1/10    Loss: 4.360400251388549

Epoch:    1/10    Loss: 4.326144758224487

Epoch:    1/10    Loss: 4.283898029327393

Epoch:    1/10    Loss: 4.271939526081085

Epoch:    1/10    Loss: 4.266182660579681

Epoch:    1/10    Loss: 4.236574376106263

Epoch:    1/10    Loss: 4.1942474360466

Epoch:    1/10    Loss: 4.1914900798797605

Epoch:    1/10    Loss: 4.157561229228973

Epoch:    2/10    Loss: 4.097775180538921

Epoch:    2/10    Loss: 4.027061755180359

Epoch:    2/10    Loss: 4.031246276855469

Epoch:    2/10    Loss: 4.007847315788269

Epoch:    2/10    Loss: 3.9822275524139403

Epoch:    2/10    Loss: 4.007336030960083

Epoch:    2/10    Loss: 3.991296555042267

Epoch:    2/10    Loss: 3.9788007082939147

Epoch:    2/10    Loss: 4

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


### third experiment contnued.... more epochs.....training 5 more epochs, again I am using the Checkpoint of previously trained model (above cell-10 epochs). Here I am passing trained_rnn

In [53]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
#######################################################################################
trained_rnn = train_rnn(trained_rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 5 epoch(s)...


/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:53: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().


Epoch:    1/5     Loss: 3.3992967495918274

Epoch:    1/5     Loss: 3.418462936401367

Epoch:    1/5     Loss: 3.4096333084106445

Epoch:    1/5     Loss: 3.404696776866913

Epoch:    1/5     Loss: 3.4237216506004335

Epoch:    1/5     Loss: 3.397377995967865

Epoch:    1/5     Loss: 3.410610272407532

Epoch:    1/5     Loss: 3.3835905556678774

Epoch:    1/5     Loss: 3.391276352882385

Epoch:    1/5     Loss: 3.4002190117835998

Epoch:    1/5     Loss: 3.41909920167923

Epoch:    1/5     Loss: 3.411190978527069

Epoch:    1/5     Loss: 3.4141451106071474

Epoch:    1/5     Loss: 3.4145865445137025

Epoch:    2/5     Loss: 3.4263007222202146

Epoch:    2/5     Loss: 3.406217842578888

Epoch:    2/5     Loss: 3.4261413507461547

Epoch:    2/5     Loss: 3.4111416149139404

Epoch:    2/5     Loss: 3.4100811762809755

Epoch:    2/5     Loss: 3.4328836612701417

Epoch:    2/5     Loss: 3.409709668636322

Epoch:    2/5     Loss: 3.392460692882538

Epoch:    2/5     Loss: 3.420439595222473



/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** (Write answer, here)

I experiment the training 3 times with different set of hyperparameters and came with these parametric values:

- `Sequence length`: I choose to use the sequence length of 10 as I find in many blogs that for sequential tasks, this number works better in practical. 


- `Batch size`: I tried using batch size of 128 initially as in lecture also it was illustrated 128, 256 batch size is pretty good as far as we have computational power. But I didn't get better result with this value so I used later 120 in later experiment which was final hyperparameter value for this project, and with other hyperparameters this was doing good in loss convergence.


- `Number of Epochs`: First I tried 15 epochs but with other hyperparameters, the loss was not converging, then I experimented in next attempt using 10 epochs. I find it was converging well, so I loaded the checkpointed version of 10 epochs model and again trained for 5 more epochs. So in total, 15 epochs were done with final hyperparameters alone.


- `Learning rate`: I didn't experimented much with this, simply followed standard practice of 0.001 value.


- `vocab_size` and `output_size`: It was simply the length of my vocabulary that I got from the length of dictionary of vocab to int which was saved earlier in the data processing stage.


- `Embedding Dimension`: Initially I tried using 300, but the loss was baraely converging so I tried anther value of 150, but this time, with other combinations of hyperparameters I was not getting good results. So finally in last experiment I used 256 value. I find this in various implementations, few in Tensorflow examples, and few in blogs. So I picked this value, as even in leture it was illustrated that value between 200 and 400 are good for embedding dimensions  in practice.


- `Hidden Dimension`: I decided to use 512, burrowed the idea from RNN lesson.


- `Number of RNN Layers`: As in practice, number of layers for RNN works best with value 1, 2 or 3. So I experimented both 2 and 3. And in my final hyperparameters I picked the value 2 as number of layers.


- `show_every_n_batches`: I picked 500 as it is going to iterate for many batches in single epochs also, since the vocab size is large enough so I thought of using 500 to show training process every 500 times.

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [50]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [54]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param rnn: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [74]:
# run the cell multiple times to get different results!
gen_length = 300 # modify the length to your preference
prime_word = 'george' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:53: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().


george: i was just a little nervous.

jerry:(to kramer) oh, i don't want to hear. i think we could get it, i don't want to hear.

george: i know...

newman: what?

jerry: i don't know.

jerry: i don't want to hear it.

george: i know.

jerry: what?

george:(smiling) well, it's not fair.

jerry: well, i don't think so.

elaine: oh, no, not a good idea!

jerry:(to jerry) what? what?

kramer: oh... i was in the hospital and i was in the bathroom. i was wondering if you want to be a good one.

jerry:(pointing) i can't believe i'm not gonna get a little problem with him.

george: you know what?

jerry: i don't know.

jerry: you got the money.

george: no....

jerry:(quietly) oh my god.

kramer: hey buddy.

george: hey

[setting: george's apartment]

kramer: hey, jerry, i got a big deal on my head. i was thinking.(to george) you know what you mean?

kramer: oh yeah. i think we should go.

jerry:(pointing) i think that's what i said. you know what i do? you can't do it, but i don't know how y

In [75]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.